### ISSUE TITLE
Doxygen hang in parallel operation

### ISSUE THREAD

**Issue by fuchsth** [doxygen/doxygen#10251]
**Describe the bug**
Doxygen hangs idle if parallel dot and num proc is selected.

**Expected behavior**
Should run to the end using mulitple processes.

**Screenshots**
Simply stops any output on the console

**To Reproduce**
In the attached zip package the necessary setup is done using FreeRTOS software. Necessary doxyfiles are included.

**Version**
$ doxygen --version
1.9.7 (09c5d2e6bdd3bd0ca5be4e4b04cd62681f0276e7)

$ dot --version
dot - graphviz version 8.0.5 (20230508.1809)

Used is the Cygwin version of doxygen and dot.

**Stack trace**
Not yet any own build of doxygen done.

**Additional context**
Problem is also seen with the Windows version. Please check if you can reproduce it. If not I will provide logs and try to build it with debug mode.

[doxygen.zip](https://github.com/doxygen/doxygen/files/12397066/doxygen.zip)

**Comment by albert-github** [comment]
Where does it hang, I assume in the generation of the graphs. I have similar problems on Cygwin and by setting:
```
DOT_NUM_THREADS        = 1
```
the problem can be overcome. It looks like a Cygwin specific problem. I've been looking into it already but didn't find the cause of the problem (might be a threading issue under Cygwin).

**Comment by fuchsth** [comment]
Yes, it is due to dot graphics. I have checked the same but Windows versions and it worked without problem. I gave it also at try with the 22.08.2023 version of the Cygwin. Problem ist still there.
Setting the DOT_NUM_THREADS to 1 solves the issue of hanging. However, we have somewhat large projects with a good amount of dot graphics. Working with only one thread slows down the generation projects significantly.

**Comment by albert-github** [comment]
`DOT_NUM_THREADS = 1` indeed is initially quite slow, but when using the old (output) directory again it normally should not try to create that many images again (or do you build each time on an empty / new html etc. directory?)

**Comment by BrianInglis** [comment]
[Gitlab #101](https://gitlab.com/cmocka/cmocka/-/issues/101)
I just built a new copy of *Cygwin* `git pull` to test unrelated patches I'm submitting, 
but there are no changes in the issue or the garbled backtrace when attached by `gdb`. 

Are there any package test cases which specifically exercise this issue? 
Or do you have any suggestions for trimming the `doxyfile` of everything but the bits running `dot`? 

If I can point the package maintainer, or one of the Cygwin console/process developers that's good at this stuff, to a package test case, we may be able to diagnose the issue, find the original cause, and fix it.

**Comment by albert-github** [comment]
@BrianInglis 
The problem looks like to be related to the way Cygwin calls the underlying Windows system in respect to the use of the doxygen "system" call (see portable.cpp). The problem does not occur with other *nix versions nor on native the Windows version.
(I think I once heard of problems with mingw but this is of course also version on top of windows.)

**Comment by BrianInglis** [comment]
Issues often arise from treating *Cygwin* as some kind of *Windows*, rather than as a *POSIX* `newlib` based *FreeBSD*, with former *Solaris*, and now *Linux* and `glibc` compatible interfaces added, and *Windows* `WSL` features enabled in some cases, like native hard- and sym-links, and case-sensitive directories (now automatic for package *builds* if supported on the underlying filesystem, like recent *NTFS* since ~2018-02). 

I will try removing all the `__CYGWIN__` conditions from `portable.cpp` and rebuild. 
If `fork`, `spawn...`, `exec`, or `system` interfaces are used with *Unix* style args or commands, *Cygwin* makes sure to get all the required semantics correct. 

Are there any other places where *Doxygen* config might incorrectly infer *Win32* rather than *Unix*? 

For example, I always have *Mingw64* toolchain and libraries installed as they are required to rebuild *Cygwin* to test low level `newlib` and *Cygwin* interfaces patches, as `libc`, `libm`, and `syscall` interface functions requiring emulation like `fork` and low level networking are built into that `dll`, to allow *X Window* managers and GUI apps, and local and remote `ssh` and `syslog` daemons and clients to work transparently. [see comment -2](https://github.com/doxygen/doxygen/issues/10251#issuecomment-2239548937).

If anything *Windows*-like is injected into a process tree, you are getting into a very *fuzzy* area, where *Cygwin* is always trying to stop the latest *Windows* program insanities from messing up one or other process, especially if you end up trying to run *Cygwin* programs from *Windows* programs launched from *Cygwin* programs, as the *Cygwin* child ends up in a different process tree from the parent, and *Cygwin* detection of *Windows* child status may be problematic.

**Comment by albert-github** [comment]
I think that in portable.cpp there has already been made quite a good separation between `WIN32` and `__CYGWIN__`. The intention is clearly to run in case of Cygwin *nix like code.

> Are there any other places where Doxygen config might incorrectly infer Win32 rather than Unix?

This should all be located in portable.cpp /portable.h Theer are a few references in util.cpp and doxygen.cpp but they should not cause problems.


I'm looking very much forward to the results of your experiments.

**Comment by BrianInglis** [comment]
Tried with [`sqlite3.c` patch](https://github.com/user-attachments/files/16323112/doxygen-1.11-deps-sqlite3-sqlite3-c.patch) `#if __CYGWIN__` -> `#if 0 /* __CYGWIN__ */` as `portable.cpp` looked good, and without, generating *Doxygen*'s own docs as part of the build, and always hangs, in fact, using `cmake -DDOXYGEN_DOT_NUM_THREADS=1 DOT_NUM_THREADS=1` 
gave
```
22:19:47.542882 00:00:00.188888	CMake Warning:
22:19:47.542971 00:00:00.000089	  Manually-specified variables were not used by the project:
22:19:47.543080 00:00:00.000021	    DOT_NUM_THREADS
22:19:47.543101 00:00:00.000020	    DOXYGEN_DOT_NUM_THREADS
```
and `make -j1` did not allow `doxygen diagrams.cfg` to progress:
```
19:36:49.707997 00:00:00.701089	[ 81%] Generating ../html/examples/diagrams/html/index.html, ../latex/examples/diagrams/latex/refman_doc.tex
22:09:50.138175 02:33:00.430178	Subprocess killed
22:09:50.151516 00:00:00.013340	make[3]: *** [examples/CMakeFiles/examples.dir/build.make:505: html/examples/diagrams/html/index.html] Error 1
22:09:50.152292 00:00:00.000777	make[3]: *** Deleting file 'html/examples/diagrams/html/index.html'
22:09:50.167849 00:00:00.015557	make[2]: *** [CMakeFiles/Makefile2:864: examples/CMakeFiles/examples.dir/all] Error 2
22:09:50.183351 00:00:00.015501	make[1]: *** [CMakeFiles/Makefile2:845: doc/CMakeFiles/docs.dir/rule] Error 2
22:09:50.194303 00:00:00.010952	make: *** [Makefile:408: docs] Error 2
22:09:50.198176 00:00:00.003873	[1;31m*** ERROR:[0;0m make failed
```
However, `ninja docs` failed: 
```
23:01:07.064960 00:00:00.418170	[129/276] Generating ../html/examples/example/html/index.html, ../latex/examples/example/latex/refman_doc.tex
23:01:07.065026 00:00:00.000065	FAILED: html/examples/example/html/index.html latex/examples/example/latex/refman_doc.tex /usr/src/doxygen/doxygen-1.11.0-1.x86_64/build/html/examples/example/html/index.html /usr/src/doxygen/doxygen-1.11.0-1.x86_64/build/latex/examples/example/latex/refman_doc.tex 
23:01:07.065101 00:00:00.000075	/bin/sh examples/CMakeFiles/index.html-b95b574.sh 0edfe4b6dd9fbd12
23:01:07.065145 00:00:00.000044	error: configuration file example.cfg not found!
23:01:07.065170 00:00:00.000025	Doxygen version 1.11.0 (28ba8143ecb6fc4ce0f0218e36bf33b51b57c2c2*)
23:01:07.065206 00:00:00.000036	Copyright Dimitri van Heesch 1997-2021
...
```
Could this be an issue with an OOT build? 

What is the best approach to make a verbose *STC* to get maximum diagnostics to debug running `doxygen diagrams.cfg` and dig down if there are any issues running dot?

**Comment by albert-github** [comment]
I see you mention a number of problems:
- sqlite3, I think this has nothing to do with the original multi threaded problem, but there might be some problems regarding Cygwin there. As it is added just for convenience and the original source is external (https://www.sqlite.org) I think that is is better to report problems regarding sqlite3 in their bug report tool (see https://www.sqlite.org/src/wiki?name=Bug+Reports).
- `cmake -DDOXYGEN_DOT_NUM_THREADS=1 DOT_NUM_THREADS=1` Is this for the the doxygen cmake of the doxygen project? As when this is the case this doesn't work as doxygen does not use this mechanism but uses the `Doxyfil`s etc. that are in the repository.
- `make -j1` In the documentation `Doxyfile` there is the line `DOT_NUM_THREADS        = 0` meaning "unlimited" number of threads for the dot part of the processing. In configimpl.l we see in case `DOT_NUM_THREADS` is `0` the code `dotNumThreads=std::max(2u,std::thread::hardware_concurrency()+1);` is executed and on my system the `std::thread::hardware_concurrency()` returns the number of processors and not the number as specified with `make -j1`. So here we look like to have 2 problems:
  - the problem of this issue (multi threads don't work under Cygwin)
  - doxygen does not take the `-j` of the `make` command into account. This looks like logical as only 1 version of doxygen is running, so obeying the `-j1` but internally multiple "threads" are spawned,
- `ninja docs` this can have to do with the OOT build, and this looks a bit like a ninja specific problem (I'm not a ninja user).
